packages/kbn-pm/src/utils/parallelize.ts

```diff
@@ -1,12 +1,49 @@
 export async function parallelizeBatches<T>(
-  batches: Array<T[]>,
+  batches: T[][],
   fn: (item: T) => Promise<void>
 ) {
   for (const batch of batches) {
-    const running = batch.map(obj => fn(obj));
-
     // We need to make sure the entire batch has completed before we can move on
     // to the next batch
-    await Promise.all(running);
+    await parallellize(batch, fn);
   }
 }
+
+export function parallellize<T>(
+  items: T[],
+  fn: (item: T) => Promise<void>,
+  concurrency = 4
+) {
+  return new Promise((resolve, reject) => {
+    if (items.length === 0) {
+      resolve();
+      return;
+    }
+
+    let activePromises = 0;
+    const values = items.slice(0);
+
+    async function scheduleItem(item: T) {
+      activePromises++;
+
+      try {
+        await fn(item);
+
+        activePromises--;
+
+        if (values.length > 0) {
+          // We have more work to do, so we schedule the next promise
+          scheduleItem(values.shift()!);
```

Comment:
__optional nit:__ to get rid of `!`, we can also do something like this (unless we want to support `undefined` for `T`), but up to you:

```ts
 const nextValue = values.shift();
 if (nextValue !== undefined) {
```